python/pyspark/sql/observation.py

```diff
@@ -0,0 +1,147 @@
+      the Dataset during the action.
+    - It will report the value of the defined aggregate columns as soon as we reach the end of
+      the action.
+
+    The metrics columns must either contain a literal (e.g. lit(42)), or should contain one or
+    more aggregate functions (e.g. sum(a) or sum(a + b) + avg(c) - lit(1)). Expressions that
+    contain references to the input Dataset's columns must always be wrapped in an aggregate
+    function.
+
+    An Observation instance collects the metrics while the first action is executed. Subsequent
+    actions do not modify the metrics returned by `Observation.get`. Retrieval of the metric via
+    `Observation.get` blocks until the first action has finished and metrics become available.
+
+    This class does not support streaming datasets.
+
+    Examples
+    --------
+    >>> from pyspark.sql.functions import col, count, lit, max
+    >>> from pyspark.sql.observation import Observation
+    >>> observation = Observation("my_metrics")
+    >>> observed_df = df.observe(observation, count(lit(1)).alias("count"), max(col("age")))
+    >>> observed_df.count()
+    2
+    >>> observation.get
+    Row(count=2, max(age)=5)
+    """
+    def __init__(self, name=None):
+        """Constructs a named or unnamed Observation instance.
+
+        Parameters
+        ----------
+        name : str, optional
+            default is a random UUID string. This is the name of the Observation and the metric.
+        """
+        assert isinstance(name, basestring), "name should be a string"
+        self._name = name or str(uuid4())
+        self._jvm = None
+        self._jo = None
+
+    def _on(self, df, *exprs):
+        """Attaches this observation to the given :class:`DataFrame` to observe aggregations.
+
+        Parameters
+        ----------
+        df : :class:`DataFrame`
+            the :class:`DataFrame` to be observed
+        exprs : list of :class:`Column`
+            column expressions (:class:`Column`).
+
+        Returns
+        -------
+        :class:`DataFrame`
+            the observed :class:`DataFrame`.
+        """
+        assert exprs, "exprs should not be empty"
+        assert all(isinstance(c, Column) for c in exprs), "all exprs should be Column"
+        assert self._jo is None, "an Observation can be used with a DataFrame only once"
+
+        self._jvm = df._sc._jvm
+        self._jo = self._jvm.org.apache.spark.sql.Observation(self._name)
+        observed_df = self._jo.on(df._jdf,
+                                  exprs[0]._jc,
+                                  column._to_seq(df._sc, [c._jc for c in exprs[1:]]))
+        return DataFrame(observed_df, df.sql_ctx)
+
+    @property
+    def get(self):
+        """Get the observed metrics.
+
+        Waits until the observed dataset finishes its first action. Only the result of the
+        first action is available. Subsequent actions do not modify the result.
+
+        Returns
+        -------
+        :class:`Row`
```

Comment:
@cloud-fan @EnricoMi @hvanhovell, shall we return a Map instead?

Reply:
I think if we move this to a Map, than it should be the same in Scala and Java. And there should be a getAsJavaMap then to avoid something like this in Java:

    JavaConverters.mapAsJavaMap(namedObservation.get())

Reply:
Yeah, I think we might need to just return a map. I'll merge this PR as is for now but would you mind creating a followup PR to return a map please?

Reply:
Yes, this would touch unrelated code and with a followup PR we can keep this possibly controversial change separated. Will do.